### PR TITLE
Issue 494: Upgrading Postgres 9.6 -> 15

### PR DESCRIPTION
**Pull request by Hzaakk**
Upgraded postgres version to 15.
Removed all hardcoded python and postgres versions in ansible playbook with variables
Some QOL stuff added to .bashrc by ansible

Notes (I have no idea why the following 2 are the case):

The 'Upgrade pip' task isn't needed with python 3.8

When upgrading to python 3.8 via SCL, the following has to be added to the 'install django application dependencies' task:
`environment: # needed as pg_config isn't in PATH by default
    PATH: "{{ ansible_env.PATH }}:/usr/pgsql-{{ postgres_version }}/bin"`

@matthew-li also made the following changes:
- Incorporated @Hzaakk's Postgres-related changes from #493.
- Removed Ansible tasks to change the Postgres authentication method from `ident` to `md5`, since Postgres now uses `scram-sha-256` by default.
    - Note: When upgrading a 9.6 deployment to 15, any Postgres user passwords need to be reset so that they are re-encrypted using the new algorithm.
- Upgraded `psycopg2-binary` from `2.8.3` to `2.9.5`.

### CODE REVIEW

File: bootstrap/ansible/playbook.yml [hunk 1]
```diff
@@ -390,10 +400,18 @@
...
-        - meta: noop
-      # Uncomment when tasks are added.
-      # when: provisioning_tasks == true and deployment == "dev"
+
+        - name: Install vim
+          yum:
+            name: vim
+            state: present
+
+        - name: Modify .bashrc to start in application directory by default
```
Comment (matthew-li) [inline bootstrap/ansible/playbook.yml:409]: Given that this is here, you can remove the recommendation in `README.md` to manually add it to the file.

**Review by matthew-li — COMMENTED** [review]
Thanks for working on this. The code changes look good to me, and I was able to successfully provision a new VM with Postgres 10.

Two things:

1. I ran into the same `pg_restore` errors you mentioned on Slack. Documenting them here:

    ```
    pg_restore: [archiver (db)] Error while PROCESSING TOC:
    pg_restore: [archiver (db)] Error from TOC entry 4036; 2606 31349 CONSTRAINT statistics_projecttransaction statistics_projecttransaction_pkey admin
    pg_restore: [archiver (db)] could not execute query: ERROR:  could not create unique index "statistics_projecttransaction_pkey"
    DETAIL:  Key (id)=(1) is duplicated.
        Command was: ALTER TABLE ONLY public.statistics_projecttransaction
        ADD CONSTRAINT statistics_projecttransaction_pkey PRIMARY KEY (id);



    pg_restore: [archiver (db)] Error from TOC entry 4033; 2606 31341 CONSTRAINT statistics_projectusertransaction statistics_projectusertransaction_pkey admin
    pg_restore: [archiver (db)] could not execute query: ERROR:  could not create unique index "statistics_projectusertransaction_pkey"
    DETAIL:  Key (id)=(1) is duplicated.
        Command was: ALTER TABLE ONLY public.statistics_projectusertransaction
        ADD CONSTRAINT statistics_projectusertransaction_pkey PRIMARY KEY (id);



    WARNING: errors ignored on restore: 2
    ```

    Looking further into it, I noticed some discrepancies between the counts of `ProjectTransaction` and `ProjectUserTransaction` instances on production vs. on the VM after loading in this morning's backup. There are some cases where multiple transaction objects share the same ID, and maybe other issues. See [transaction_discrepancies.txt](https://github.com/ucb-rit/coldfront/files/10449424/transaction_discrepancies.txt).

    Can you look into this? (Try loading in the latest backup in a VM running 9.6 vs. a VM running 10 and see what the differences are in the two tables?)

2. Can you stake out what an upgrade of a deployment already running 9.6 would look like? I'm hoping we can leverage [`pg_upgrade`](https://www.postgresql.org/docs/current/pgupgrade.html) for this. Savio is having a downtime in a few weeks, which might be a good opportunity to do this on production.

**Comment by Hzaakk** [comment]
@matthew-li I think SCL was only needed to upgrade python to 3.8 as there weren't any other repos available for RHEL 7. Is there something else for PSQL it's used for that I'm missing?

**Comment by matthew-li** [comment]
> @matthew-li I think SCL was only needed to upgrade python to 3.8 as there weren't any other repos available for RHEL 7. Is there something else for PSQL it's used for that I'm missing?

Yeah, I was getting something about “requires llvm-toolset-clang-7” when trying to install `postgresql15-devel` until I added SCL.

**Comment by matthew-li** [comment]
Below are some rough notes on using `pg_upgrade` to upgrade an existing 9.6 instance to 15. I'm leaving these here for now, but will incorporate them into long-term documentation elsewhere.

Any existing dev VMs running 9.6 can use these to upgrade in place, without having to dump/reload the database.

```
# Install PostgreSQL 15.
# Change gather_facts to true in the playbook.
# Add the following task to the end of the playbook.

#     - name: Install SCL
#       include_role:
#         role: smbambling.scl

# Run the task in the playbook.
ansible-playbook bootstrap/ansible/playbook.yml --start-at-task "Install SCL"

# Install via yum.
sudo yum install postgresql15
sudo yum install postgresql15-libs
sudo yum install postgresql15-devel
sudo yum install postgresql15-test
sudo yum install postgresql15-server


# Initialize, enable, and start the database.
sudo /usr/pgsql-15/bin/postgresql-15-setup initdb
sudo systemctl enable postgresql-15
sudo systemctl start postgresql-15


# Stop both PostgreSQL instances.
sudo systemctl stop postgresql-9.6
sudo systemctl stop postgresql-15


# Run pg_upgrade to go from 9.6 to 15.
sudo su - postgres
/usr/pgsql-15/bin/pg_upgrade --old-datadir /var/lib/pgsql/9.6/data --new-datadir /var/lib/pgsql/15/data --old-bindir /usr/pgsql-9.6/bin --new-bindir /usr/pgsql-15/bin


# Start PostgreSQL 15.
sudo systemctl start postgresql-15


# Change the admin user password so that it is encrypted with scram-sha-256, not md5.
# https://stackoverflow.com/a/12721095
su - postgres
psql cf_brc_db  # May be cf_lrc_db
> ALTER USER admin WITH PASSWORD 'YOURPASS';


# Test that authentication works.
PGPASSWORD=YOURPASS psql -h localhost -d cf_brc_db -U admin    # May be cf_lrc_db
```

**Comment by matthew-li** [comment]
Thanks again for working through the different versions and re-provisioning repeatedly.

I've provisioned two VMs for BRC and LRC. When loading the latest database backups for both, I did not encounter the `pg_restore` issue noted above. Everything appears to be in order.

On staging and production, I'll run `pg_upgrade`, as noted above.

The failing test is intermittent, and related to the timestamp comparison you'd been updating in #493.

Merging.

**Pull request merged**